src/tensor_ops/matmul.rs

```diff
@@ -159,6 +159,106 @@ pub fn vecmat_mul_transpose<const K: usize, const N: usize, TAPE: Tape>(
     })
 }
 
+/// Batch matrix multiplication.
+///
+/// # Generics
+/// - `B`: Batch size in `lhs`.
+/// - `M`: number of rows of `lhs`.
+/// - `K`: number of columns of `lhs` and number of rows of `rhs`.
+/// - `N`: Number of columns of `rhs`.
+///
+/// # Arguments
+/// * `lhs` - a 3d tensor representing a BxMxK matrix
+/// * `rhs` - a 2d tensor representing a KxN matrix
+///
+/// Returns a 3d tensor representing a BxMxN matrix.
+///
+/// # Examples
+///
+/// ```rust
+/// # use dfdx::prelude::*;
+/// let x: Tensor3D<5, 3, 2> = Tensor3D::zeros();
+/// let y: Tensor2D<2, 4> = Tensor2D::zeros();
+/// let result: Tensor3D<5, 3, 4> = batch_matmul(x, &y);
+/// ```
+pub fn batch_matmul<const M: usize, const K: usize, const N: usize, const B: usize, TAPE: Tape>(
+    lhs: Tensor3D<B, M, K, TAPE>,
+    rhs: &Tensor2D<K, N, NoneTape>,
+) -> Tensor3D<B, M, N, TAPE> {
+    let mut result = Tensor3D::zeros();
+    bmm(lhs.data(), rhs.data(), result.mut_data());
+
+    // copy rhs data for use later when computing gradients
+    let rhs_data = rhs.data.clone();
+
+    move_tape_and_add_backward_binop(lhs, rhs, result, move |lhs, rhs, result, grads| {
+        let (lhs_grad, result_grad) = grads.mut_and_ref(&lhs, &result);
+        bmm_bt(result_grad, rhs_data.as_ref(), lhs_grad);
+
+        let (rhs_grad, result_grad): (&mut [[f32; N]; K], &[[[f32; N]; M]; B]) =
+            grads.mut_and_ref(&rhs, &result);
```

Comment:
Yeah I could not figure out why you need to type hint for these.

Reply:
Couldn't either lol. Spent a while figuring out I just needed those. Probably some compiler bug.